src/material/datepicker/month-view.ts

```diff
@@ -329,6 +340,9 @@ export class MatMonthView<D> implements AfterContentInit, OnChanges, OnDestroy {
       this.activeDateChange.emit(this.activeDate);
     }
 
+    // Ensure the calendar body has the correct active cell.
+    this._changeDetectorRef.detectChanges();
```

Comment:
Why is this necessary? This would invoke an entire extra change detection, which we generally want to avoid.

Reply:
Yeah, this seems a little iffy. I'm guessing that it might be flushing this subscription: https://github.com/angular/components/blob/master/src/material/datepicker/calendar-body.ts#L201

Reply:
After debugging this with console.log statements, I determined that the calendar body calls .focus on the active cell before receiving an updated value for activeCell. This causes it to focus on the old active cell, rather than the one set in the keydown handler.

Ensuring that change detection runs appears to be the fix here, but I'm curious why this wan't a problem until now.

Reply:
If it works in real life but not in tests, it most likely means the _test_ needs a `detectChanges` call and not the source code (which is very normal since in the test nothing is causing change dtections to happen autmatically)

Reply:
I just pushed a change to only call `detectChanges` if the event is not trusted. This is only needed for the tests because fake keyboard events do not trigger change detection. I'm generally not a fan of having code that's only intended to run in the tests, and I'm curious how we usually handle this situation.

Calling detectChanges in the test code doesn't work because by that time the calendar body will have already triggered the `activeDateChange` event with a stale value. This wasn't a problem until now, because before focusing on a stale date wouldn't trigger the `activeDateChange` event to update the active date. Calling detectChanges in the test code would case the calendar body to focus again but on a updated value.

Here's a rough timeline of what happens in the unit tests if we don't call `detectChanges` when handling the `keydown` event.
0. (active Date is Jan 12 2021)
1. trigger a fake keyboard event with the right arrow key
2. MatMonthView sets `this.activeDate` to Jan 13 2021
3. MatMonthView calls `_focusActiveDate`
4. MatCalendarBody calls .focus on Jan 12, 2021 date then triggers an `activeDateChange` Event with Jan 12, 2021
5. test code calls `detectChanges`
6. month-view receives sets the active date to Jan 12, 2021
7. (active date should be Jan 13, 2021).

Reply:
Is there a supported way to do this programmatically without dispatching fake events? If there isn't, teams will probably try dispatching keydown events manually to get this desired behavior. So if the support for this gets baked into the datepicker it will likely be depended on

Reply:
> I'm generally not a fan of having code that's only intended to run in the tests, and I'm curious how we usually handle this situation.

This isn't really the way to think about this situation. In a real application, Angular will trigger change detection when a user interacts with the page (clicks, keypresses, etc). However, that won't happen if we trigger a fake event in a test (since obviously you can't trigger a real event in a test). So calling `detectChanges` in the test is the thing that most closely approximates what happens when the real app is running. The library code should almost _never_ have unnecessary behaviors added to support tests (barring things like extra attrs we add for harnesses).

Reply:
Okay, so I should think about if the behavior is different between the tests and a real application, rather than the exact code path.

> The library code should almost never have unnecessary behaviors added to support tests

Does the library code mean Angular Material (this repo) or the Angular framework?

The above diff is out-of-date, but I made it only call `detectChanges` if the event is not trusted, so that we don't impact the performance in a real application.